File: blackjack.cpp
// TODO: read the following
// Use the following
//    g++ -g -std=c++11 blackjack.cpp
//
// TODO: add a remove for all news
#include <iostream>
#include <iterator>
#include <string>
#include <list>
#include <algorithm>

using namespace std;

class Card{
	protected:
		string suit; // clubs, hearts, diamonds, spades
		int value;
		int altValue;
		string cardName;
	public:
		Card(string suit, string cardName){
			this->suit = suit;
			this->cardName = cardName;
			if(cardName == "J" || cardName == "Q" || cardName == "K"){
				value = 10;
				altValue = 10;
			}
			else if(cardName == "A"){
				value = 11;
				altValue = 1;
			}
			else{
				value = stoi(cardName);
				altValue = stoi(cardName);
			}
		}
		string getSuit(){return suit;}
		int getValue(){return value;}
		int getAltValue(){return altValue;}
		string getCardName(){return cardName;}
		string printCard(){return cardName + suit;}
		bool isSame(Card card){
			if(card.getSuit() == this->getSuit() && card.getCardName() == this->getCardName())
				return true;
			return false;
		}
		~Card(){}

};

// Pile cannot be instantiated
class Pile{
	protected:
		list<Card> cards;
	public:
		void addCard(Card card){
			cards.push_back(card);
		}
		Card removeCard(Card card){
			for(list<Card>::iterator card_i = cards.begin(); card_i != cards.end(); ++card_i){
				if(card.isSame(*card_i)){
					Card toDelete = *card_i;
					cards.erase(card_i);
					return toDelete;
				}
			}
		}
		~Pile(){}
};

class Deck: public Pile{
	protected:
		const static int suitNum = 4, cardTypeNum = 13;
		const string heart = "\xe2\x99\xa5";//https://www.fileformat.info/info/unicode/char/2665/index.htm
		const string spade = "\xe2\x99\xa0";//https://www.fileformat.info/info/unicode/char/2660/index.htm
		const string diamond = "\xe2\x99\xa6";//https://www.fileformat.info/info/unicode/char/2666/index.htm
		const string club = "\xe2\x99\xa3";//https://www.fileformat.info/info/unicode/char/2663/index.htm
		string suits[suitNum] = {club, heart, diamond, spade};
		string cardNames[cardTypeNum] = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};
	public:
		// Constructor creates a complete card deck (no jokers)
		Deck(){
			for(int i = 0; i < suitNum; i++){
				for(int j = 0; i < cardTypeNum; j++){
					Card *card = new Card(suits[i], cardNames[j]);
					this->addCard(*card);
				}
			}
		}
		void shuffle(){
			random_shuffle(cards.begin(), cards.end());
		}
		Card draw(){
			list<Card>::iterator toDelete = cards.begin();
			cards.erase(toDelete);
			return *toDelete;
		}
		void deal(list<Pile>& activePlayers){
			int size = activePlayers.size();
			for(int i = 0; i < 2; i++){
				for(list<Pile>::iterator activePlayers_i = activePlayers.begin(); activePlayers_i != activePlayers.end(); ++activePlayers_i){
					activePlayers_i->addCard(draw());
				}
			}
			return;
		}
		void deal(Pile& dealer){
			for(int i=0;i<2;i++){
				dealer.addCard(draw());
			}
			return;
		}
		~Deck(){}
};

class Player: public Pile{
	private:
		static Deck deck;
		int currentPoints;
	public:
		int getCurrentPoints() const{return currentPoints;}
		int calculateCurrentPoints(){
			int value = 0;
			int altValue = 0;
			for(auto card = cards.begin(); card != cards.end(); ++card){
				value += card->getValue();
				altValue += card->getAltValue();
			}
			if(value <= 21){
				currentPoints = value;
				return value;
			}
			else{
				currentPoints = altValue;
				return altValue;
			}
		}
		inline bool operator <(const Player &player) const{
			return this->getCurrentPoints() < player.getCurrentPoints();
		}
		bool isBust(){
			if(this->calculateCurrentPoints() > 21)
				return true;
			else
				return false;
		}
		void hit(){
			this->addCard(deck.draw());
			this->calculateCurrentPoints();
		}
};

class Dealer: public Player{
	public:
        bool play(int scoreToBeat){
			if(isBust())
				return false;
			else if(calculateCurrentPoints()>=scoreToBeat)
				return true;
			else{
				this->hit();
				return this->play(scoreToBeat);
			}
        }
		string showPublicHand(){
			list<Card>::iterator toShow = cards.begin();
			return toShow->printCard() + " ??\n";
		}
};

class User: public Player{
	private:
		string name;
		int chips = 500;
	public:
		User(string name){
			this->name = name;
		}
		string getName(){return name;}
		double getChips(){return chips;}
		void setName(string name){this->name = name;}
		string getHand(){
			string hand = "";
			for(auto card = cards.begin(); card != cards.end(); ++card)
				hand += card->printCard() + " ";
			hand += "\n";
			return hand;
		}
        bool play(string dealerHand){
			if (isBust())
				return false;
			else if (calculateCurrentPoints() == 21)
				return true;
			else{
				while(true){
					string moveMessage = string("Dealer: " + dealerHand + " " + this->getName() + ": " + this->getHand() + "\nCurrent point count: " + string(this->calculateCurrentPoints()) + "\nHit? (y/n): ");
					cout << moveMessage;
					char user;
					cin >> user;
					if(user == 'y' || user == 'Y'){
						this->hit();
						return this->play(dealerHand);
					}
					else if(user == 'n' || user == 'N'){
						return true;
					}
					else
						cout << "Error! Invalid input.\n";
				}
			}
        }
};

class Game{
	protected:
		// https://www.geeksforgeeks.org/list-cpp-stl/
		list<Player> activePlayers;
		Dealer *dealer;
		int rounds;
		Deck *deck;
	public:
		Game(int playerNumber){
			Deck *deck = new Deck();
			this->deck = deck;
			cout << "Welcome to blackjack\n";
			Dealer *dealer = new Dealer();
			this->dealer = dealer;
			this->addPlayer(this->dealer); // Add dealer in first location
			playerNumber > 5 ? playerNumber = 5 : playerNumber = playerNumber; // Set maximum opponents to be 5
			playerNumber < 1 ? playerNumber = 1 : playerNumber = playerNumber; // Set minimum opponents to be 1
			string name;
			for(int i = 0; i < playerNumber; i++){
				cout << "Enter name: ";
				cin >> name;
				User *user = new User(name);
				this->addPlayer(user);
			}
		}
		int getRounds(){return rounds;}
		bool addPlayer(Player &player){activePlayers.push_back(player);}
		void printPlayers(){ // Print all player names and chip count
			// https://stackoverflow.com/questions/22269435/how-to-iterate-through-a-list-of-objects-in-c/22269465
			for(list<User>::iterator user = activePlayers.begin(); user != activePlayers.end(); ++user)
				cout << "Player: " << user.getName() << "  |  $" << user.getChips() << endl;
		}
		void gameSummary(){
			cout << "Total rounds: " << getRounds() << endl;
			this->printPlayers();
		}
		void checkPlayers(){ // Check if any players have run out of chips
			for(list<User>::iterator user = activePlayers.begin(); user != activePlayers.end(); ++user){
				if(player.getChips() <= 0)
					activePlayers.erase(player);
			}
		}
		int getTopcalculateCurrentPoints(){
			for(list<Player>::iterator user = activePlayers.begin(); user != activePlayers.end(); ++user){
				if(user->calculateCurrentPoints !> 21)
					return user->calculateCurrentPoints;
			}
		}
		void newRound(){
			rounds++;
			this->checkPlayers();
			deck.shuffle();
			deck.deal(activePlayers);
			deck.deal(dealer);

			// All players play
			for(list<User>::iterator user = activePlayers.begin(); user != activePlayers.end(); ++user){
				user->play(dealer.showPublicHand());
			}
			activePlayers = activePlayers.sort();
			dealer.play(getTopcalculateCurrentPoints());

			// for(list<User>::iterator user = activePlayers.begin(); user != activePlayers.end(); ++user){
			// 		if(user->isBust()){ // If user bust, they lose
						
			// 		}
			}
		}
		~Game(){

		}
};

int main(){
	Game *game = new Game(2);
	int user;
	do{
		cout << "Enter 1 to play, 2 to check score, and 0 to quit\n... ";
		cin >> user;
		switch(user){
			case 1:
				game->newRound();
				break;
			case 2:
				game->gameSummary();
				break;
			case 0:
				break;
			default:
				cout << "Error! Invalid option.\n";
		}
	}while (user != 0);
	return 0;
}

File: blackjack.py
#! /bin/python
import random

# Use for all class basics and inheritance syntax in python https://www.geeksforgeeks.org/inheritance-in-python/

# Author: Kai Arsenault and Daniel Arpide
class Card:
    def __init__(self, suit, cardName, value):
        self.suit = suit
        self.cardName = cardName
        self.value = value
        self.altValue = value
        if value == 11:
            self.altValue = 1
    def getSuit(self):
        return self.suit
    def getValue(self):
        return self.value
    def getAltValue(self):
        return self.altValue
    def getCardName(self):
        return self.cardName
    # Display card: card name followed by suit
    def getCard(self):
        return self.cardName + self.suit
    # Check if card is the same with the card being passed
    def isSame(self, card):
        if card.getSuit() == self.suit and card.getCardName() == self.cardName:
            return True
        return False

# Author: Kai Arsenault
# Pile is not made to be instantiated but is a base class for anything that uses a list of cards
class Pile:
    # Each Pile class has a list of cards
    def __init__(self):
        self.cards = []
    def addCard(self, card):
        self.cards.append(card)
    # Find if this Pile has the card passed, if it does delete it and return the deleted card
    def removeCard(self, cardToDelete):
        for i in range(len(cards)):
            if self.cards[i].isSame(cardToDelete):
                # https://www.geeksforgeeks.org/python-list-pop/
                return self.cards.pop(i)
        return None

# Author: Daniel Arpide
# Derived from Pile, is a stack of cards shared by all players, is instantiated as a full deck of cards
class Deck (Pile):
    #suits = ['♠', '♦', '♥', '♣']
    club = "\xe2\x99\xa3"
    spade = "\xe2\x99\xa0"
    heart = "\xe2\x99\xa5"
    diamond = "\xe2\x99\xa6"
    suits = [club, spade, heart, diamond]
    #Dictionary of all card values. Key - Name, Value - Points
    values = {'2' : 2, '3' : 3, '4' : 4, '5' : 5, '6' : 6, '7' : 7, '8' : 8, '9' : 9, '10' : 10, 'J' : 10, 'Q' : 10, 'K' : 10, 'A' : 11}
    # Constructor creates a full 52 card deck
    def __init__(self):
        # https://cardsoverflow.com/questions/5166473/inheritance-and-init-method-in-python
        Pile.__init__(self)
        for suit in self.suits:
            for value in self.values:
                self.addCard(Card(suit, value, self.values[value]))
    def shuffle(self):
        # https://www.geeksforgeeks.org/python-ways-to-shuffle-a-list/
        random.shuffle(self.cards)
    # Draw a card from the top of the deck
    def draw(self):
        return self.cards.pop(0)
    # Deal 2 cards to each player in the list of players passed and to the dealer
    def deal(self,players, dealer):
        for i in range(2):
            for player in players:
                player.addCard(self.draw())
            dealer.addCard(self.draw())

# Author: Kai Arsenault
# Player class derived from Pile, not meant to be instantiated, used for User and Dealer
class Player (Pile):
    deck = None
    # Any player that is created must be "tied" to a deck. Each player has a deck it uses
    def __init__(self, deck):
        Pile.__init__(self)
        self.deck = deck
    # Find players total points, if points are over 21 check the alternate points (any Ace as 1 instead of 11)
    def getPoints(self):
        value = 0
        altValue = 0
        for card in self.cards:
            value += card.getValue()
            altValue += card.getAltValue()
        if value > 21:
            return altValue
        else:
            return value
    # Return each card in player's hand/all players cards
    def displayCards(self):
        hand = ''
        for card in self.cards:
            hand += card.getCard() + ' '
        return hand
    # Checks if players points are over 21, returns true or false accordingly
    def isBust(self):
        if self.getPoints() > 21:
            return True
        return False
    # Add card to players hand from deck
    def hit(self):
        card = self.deck.draw()
        self.addCard(card)
        return card
    # Empties user's hand putting the cards back into deck
    def putCardsBack(self):
        while len(self.cards) > 0:
            self.deck.addCard(self.cards.pop())

#Author: Kai Arsenault and Noah Damergi (sort of)
class Dealer (Player):
    # Dealer must have a deck
    def __init__(self, deck):
        Player.__init__(self, deck)
    # Get dealer's current score (how many points they have in their hand)
    def getCurrentScore(self):
        return self.currentScore
    # Dealer plays based off the score passed. They will keep hitting until the same or higher than the score past or until bust
    def play(self, topScore):
        # Checks if dealer has busted
        if(self.isBust()):
            print('Bust!\n')
            self.currentScore = self.getPoints()
            print('Dealer: ' + self.displayCards() + ' Dealer total: ' + str(self.getPoints()) + '\nDealer Bust!\n')
            return self.currentScore
        # Checks if dealer has beat or is the same as the top score
        elif(self.getPoints() >= topScore):
            self.currentScore = self.getPoints()
            print('Dealer: ' + self.displayCards() + ' Dealer total: ' + str(self.getPoints()) + '\n')
            return self.currentScore
        # If dealer is not bust and is under top score will hit and reruns function
        else:
            self.hit()
            self.play(topScore)
    # Return dealers hand, only displaying the second card
    def displayCards_protected(self):
        hand = ''
        for i in range(len(self.cards) - 1):
            hand += '?? '
        # https://www.geeksforgeeks.org/python-how-to-get-the-last-element-of-list/
        hand += self.cards[-1].getCard()
        return hand

#Author: Kai Arsenault
class User (Player):
    # User must have a deck
    def __init__(self, deck, name):
        Player.__init__(self, deck)
        self.chips = 500
        self.name = name
    # Get chip total for user
    def getChips(self):
        return self.chips
    def getName(self):
        return self.name
    def changeName(self, name):
        self.name = name
    # Get card total of current hand
    def getCurrentScore(self):
        return self.currentScore
    # User loses the amount they bet
    def lostBet(self):
        self.chips -= self.bet
        print(self.getName() + ' lost ' + str(self.bet) + ' chips. New total: ' + str(self.chips))
    # User wins the amount they bet
    def wonBet(self):
        self.chips += self.bet
        print(self.getName() + ' won ' + str(self.bet) + ' chips. New total: ' + str(self.chips))
    # User can place a bet, must be greater than 0 (must bet something) and must be less than chip total
    def placeBet(self):
        while True:
            print('Current chips:', self.chips)
            self.bet = input('Enter bet amount (only dollar amount, cannot bet cents)\n... ')
            self.bet = int(self.bet)
            if self.bet <= 0:
                print('Error! Bet must be greater than 0.')
            elif self.bet > self.chips:
                print('Error! Cannot bet more than you have.')
            else:
                break
    # Users plays until bust or has 21 (perfect score). Each round user has choice of whether to hit
    def play(self):
        if self.isBust():
            print('\nBust!')
            self.currentScore = self.getPoints()
            return self.getPoints()
        elif self.getPoints() == 21:
            print('\nGot 21, highest possible score.\n' + self.name + ':', self.displayCards(), 'Current total:', self.getPoints())
            self.currentScore = self.getPoints()
            return self.getPoints()
        else:
            print('\n' + self.name + ':', self.displayCards(), 'Current total:', self.getPoints())
            while True:
                # Gives option to hit
                choice = input('Hit (y/n)? ')
                if choice == 'y' or choice == 'Y':
                    cardAdded = self.hit()
                    # Shows card that was picked up
                    print('Picked up:', cardAdded.getCard())
                    return self.play()
                elif choice == 'n' or choice == 'N':
                    self.currentScore = self.getPoints()
                    return self.getPoints()
                else:
                    print('Error! Invalid input.')

# Author: Kai Arsenault
class Game:
    # List of all users that are playing
    users = []
    # Sets rounds to 0, creates a deck, dealer, and users based on how many people are playing
    def __init__(self):
        print('Welcome to Blackjack.')
        self.rounds = 0
        self.deck = Deck()
        self.dealer = Dealer(self.deck)
        self.getUserNum()
        for i in range(int(self.userNum)):
            name = input('Enter name.\n... ')
            self.users.append(User(self.deck, name))
    # Check to see how many users are playing, must be between 1 and 6
    def getUserNum(self):
        self.userNum = input('Enter number of players (1-6 players allowed).\n... ')
        while int(self.userNum) < 1 or int(self.userNum) > 6:
            self.userNum = input('Error! Number of players must be between 1 and 6.\n... ')
    # Find how many rounds have been played
    def getRounds(self):
        return self.rounds
    # Get leaderboard, displays all players (still in the game) in order by number of chips
    def getScoreBoard(self):
        scoreBoard = '\nLeaderboard:\n'
        sortedUsers = sorted(self.users, key=lambda user: user.getChips())
        for i in range(len(sortedUsers)):
            scoreBoard += '#' + str(i + 1) + ' ' + sortedUsers[i].getName() + ' || $' + str(sortedUsers[i].getChips()) + '\n'
        return scoreBoard
    # Displays leaderboard with total rounds printed. Used at the end of the game
    def getGameSumary(self):
        gameSummary = '\n----------\nEND\n----------\nTotal rounds: ' + str(self.rounds)
        gameSummary += self.getScoreBoard()
        return gameSummary
    # Checks all users to see if they still have chips, if they are out, they are removed from game
    def checkUsers(self):
        usersToDelete = []
        for i in range(len(self.users)):
            if self.users[i].getChips() <= 0:
                usersToDelete.append(i)
        for userIndex in usersToDelete:
            self.users.pop(userIndex)
    # Simulates a round of blackjack
    def newRound(self):
        # Start of round sequence: add round, remove users out of chips, shuffle the deck, and deal cards
        self.rounds += 1
        self.checkUsers()
        self.deck.shuffle()
        self.deck.deal(self.users, self.dealer)
        # All users play: all users play, display dealers protected hand (just second card)
        for user in self.users:
            print('\n----------\n' + user.getName() + "'s turn.")
            user.placeBet()
            print('\nDealer:', self.dealer.displayCards_protected())
            user.play()
        # Get user top score: find the top score among the users that is not bust
        topScore = 0
        for user in self.users:
            if user.getCurrentScore() <= 21 and user.getCurrentScore() > topScore:
                topScore = user.getCurrentScore()
        # Dealer plays: plays based on top score
        self.dealer.play(topScore)
        # Settle bets: all users that have won or lost recieve or lose chips repectively
        for user in self.users:
            # If dealer wins
            if user.isBust() or (user.getCurrentScore() < self.dealer.getCurrentScore() and not self.dealer.isBust()):
                user.lostBet()
            else:
                user.wonBet()
        # End of round sequence: print updated score board with all players chip totals, all players put cards back
        print(self.getScoreBoard())
        for user in self.users:
            user.putCardsBack()
        self.dealer.putCardsBack()

def main():
    game = Game()
    # Player has the option to quit or to play again, will repeat until players quit
    while True:
        choice = input('\nOptions:\n  1 - Play hand\n  2 - Quit game\n... ')
        choice = int(choice)
        if choice == 1:
            game.newRound()
        elif choice == 2:
            print(game.getGameSumary())
            break
        else:
            print('Error! Invalid input.')

if __name__ == '__main__':
    main()